File: 120. Triangle/Solution.py
from typing import List

class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        dp = [[None] * n for _ in range(n)]   # memoization table
        return self.helper(triangle, 0, 0, dp)

    def helper(self, triangle: List[List[int]], row: int, col: int, dp: List[List[int]]) -> int:
        if row == len(triangle):
            return 0

        if dp[row][col] is not None:
            return dp[row][col]

        result = triangle[row][col] + min(
            self.helper(triangle, row + 1, col, dp),
            self.helper(triangle, row + 1, col + 1, dp)
        )

        dp[row][col] = result
        return result

File: 120. Triangle/Solution.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        Integer[][] dp = new Integer[201][201];
        return helper(triangle, 0, 0, dp);
    }

    private int helper(List<List<Integer>> triangle, int rowIndex, int cRowIndex, Integer[][] dp) {
        if (rowIndex == triangle.size()) {
            return 0;
        }

        if (dp[rowIndex][cRowIndex] != null) {
            return dp[rowIndex][cRowIndex];
        }

        int result = triangle.get(rowIndex).get(cRowIndex) + 
        Math.min(helper(triangle, rowIndex + 1, cRowIndex, dp), helper(triangle, rowIndex + 1, cRowIndex + 1, dp));

        dp[rowIndex][cRowIndex] = result;

        return dp[rowIndex][cRowIndex];
    }
}
